src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -0,0 +1,180 @@
+    // TODO: also needs to initialize base class blkallocator for recovery path;
+
+    m_sb.load(buf, meta_cookie);
+    m_last_append_offset = m_sb->last_append_offset;
+    m_freeable_nblks = m_sb->freeable_nblks;
+
+    HS_REL_ASSERT_EQ(m_sb->magic, append_blkalloc_sb_magic, "Invalid AppendBlkAlloc metablk, magic mismatch");
+    HS_REL_ASSERT_EQ(m_sb->version, append_blkalloc_sb_version, "Invalid version of AppendBlkAllocator metablk");
+}
+
+// alloc a single block;
+BlkAllocStatus AppendBlkAllocator::alloc(BlkId& bid) {
+    if (available_blks() < 1) {
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("No space left to serve request nblks: 1, available_blks: {}", available_blks());
+        return BlkAllocStatus::SPACE_FULL;
+    }
+
+    bid.set(m_last_append_offset, 1, m_chunk_id);
+    ++m_last_append_offset;
+    --m_freeable_nblks;
+
+    set_dirty_offset();
+
+    COUNTER_INCREMENT(m_metrics, num_alloc, 1);
+    return BlkAllocStatus::SUCCESS;
+}
+
+//
+// For append blk allocator, the assumption is only one writer will append data on one chunk.
+// If we want to change above design, we can open this api for vector allocation;
+//
+BlkAllocStatus AppendBlkAllocator::alloc(blk_count_t nblks, const blk_alloc_hints& hint,
+                                         std::vector< BlkId >& out_bids) {
+    if (available_blks() < nblks) {
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("No space left to serve request nblks: {}, available_blks: {}", nblks, available_blks());
+        return BlkAllocStatus::SPACE_FULL;
+    } else if (nblks > BlkId::max_blks_in_op()) {
+        // consumer(vdev) already handles this case.
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("Can't serve request nblks: {} larger than max_blks_in_op: {}", nblks, BlkId::max_blks_in_op());
+        return BlkAllocStatus::FAILED;
+    }
+
+    // Push 1 blk to the vector which has all the requested nblks;
+    out_bids.emplace_back(m_last_append_offset, nblks, m_chunk_id);
+
+    m_last_append_offset += nblks;
+    m_freeable_nblks -= nblks;
+
+    set_dirty_offset();
+
+    COUNTER_INCREMENT(m_metrics, num_alloc, 1);
+
+    return BlkAllocStatus::SUCCESS;
+}
+
+//
+// Alternative design option:
+// Option-1. Everytime offset is updated in a single append_blk_allocator instance, update the dirty buffer into
+// VDevCPContext, which contains a list of (ThreadVector) dirty buffers from all the append_blk_allocator instances
+// (protected by RCU);
+//
+// And cp_flush flush every dirty buffer in that list to disk;
+//
+// Option-2. vdev cp_flush will call each individual chunk's blkallocator and each append_blk_allocator instance flush
+// its dirty buffer on demond;
+//
+// We are using option-2.
+//
+
+//
+// cp_flush is happening in different thread that is running concurrently as AppendBlkAllocator::alloc;
+//
```

Comment:
yes. it is nothing different as a bitmap we are doing for varsize blk allocator, only difference is instead of a bitmap, we are persisting a few counter/offsets.